nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1872,63 +2111,132 @@ impl<'a> Planner<'a> {
+        self.blueprint
+            .parent_blueprint()
+            .all_omicron_zones(BlueprintZoneDisposition::is_in_service)
+            .find_map(|(_, blueprint_zone)| {
+                if self.input.active_nexus_zones().contains(&blueprint_zone.id)
+                {
+                    Some(blueprint_zone.image_source.clone())
+                } else {
+                    None
                 }
+            })
+    }
+
+    fn lookup_current_nexus_generation(
+        &self,
+    ) -> Result<Option<Generation>, Error> {
+        // Look up the active Nexus zone in the blueprint to get its generation
+        self.blueprint
+            .parent_blueprint()
+            .find_generation_for_nexus(self.input.active_nexus_zones())
+            .map_err(|_| Error::NoActiveNexusZonesInParentBlueprint)
+    }
 
-                Ok(true)
+    // Returns if the zone is ready to be updated.
+    //
+    // If the zone should not be updated yet, updates the planner report to
+    // identify why it is not ready for update.
+    fn can_zone_be_updated(
+        &self,
+        zone: &BlueprintZoneConfig,
+        report: &mut PlanningZoneUpdatesStepReport,
+    ) -> Result<bool, Error> {
+        let safe_to_update = self.can_zone_be_shut_down_safely(zone, report);
+
+        let zone_nexus_generation = match &zone.zone_type {
+            // For Nexus, we need to confirm that the active generation has
+            // moved beyond this zone.
+            BlueprintZoneType::Nexus(nexus_zone) => {
+                // Get the nexus_generation of the zone being considered for shutdown
+                nexus_zone.nexus_generation
             }
-            _ => Ok(true), // other zone kinds have no special dependencies
+            _ => return Ok(safe_to_update),
+        };
+
+        use ZoneWaitingToExpunge::*;
+
+        // Get the generation of the currently-executing Nexus zones.
+        //
+        // This presumably includes the currently-executing Nexus where
+        // this logic is being considered.
+        let Some(current_gen) = self.lookup_current_nexus_generation()? else {
+            // If we don't know the current Nexus zone ID, or its
+            // generation, we can't perform the handoff safety check.
+            report.waiting_zone(
+                zone,
+                Nexus {
+                    zone_generation: zone_nexus_generation,
+                    current_nexus_generation: None,
+                },
+            );
+            return Ok(false);
+        };
+
+        // We need to prevent old Nexus zones from shutting themselves
+        // down. In other words: it's only safe to shut down if handoff
+        // has occurred.
+        //
+        // That only happens when the current generation of Nexus (the
+        // one running right now) is greater than the zone we're
+        // considering expunging.
+        if current_gen <= zone_nexus_generation {
+            report.waiting_zone(
```

Comment:
Ah, now I understand what "waiting to be expunged" means...  I'm still not sure this is worth reporting though?  This is just an ordinary case of a Nexus zone running at the current generation, isn't it?

Reply:
I still think it's worth reporting in https://github.com/oxidecomputer/omicron/commit/9843b4a4066968311b38179164c115e1de26d00d, and have a test case where "a blueprint could be emitted, telling you what needs to change, precisely when this is relevant".

Hopefully tightening the window makes this more clear. I still think it's valuable for the planner to self-identify when it's "blocked/waiting" for something, which would be the case here.